File: src/clustr/tools/read/updates.py
"""
Read-only tool: is the connected Proxmox cluster running the latest release?

Reads the running version from the Proxmox API (``/version``) and compares it to
the latest Proxmox VE release published on the official roadmap
(https://pve.proxmox.com/wiki/Roadmap).

Best-effort by design: the outbound fetch to pve.proxmox.com may be blocked by a
locked-down network policy or simply offline. In that case the running version
is still reported and the upstream lookup degrades to an actionable message
rather than failing the tool. This tool never mutates the cluster, so
``readOnlyHint = True``.
"""

from __future__ import annotations

import logging
import re
from typing import Any

from mcp.server.fastmcp import FastMCP
from mcp.types import ToolAnnotations

from clustr.proxmox.client import get_client, proxmox_get
from clustr.tools import safe

logger = logging.getLogger(__name__)

_READ_ONLY = ToolAnnotations(
    readOnlyHint=True, destructiveHint=False, idempotentHint=True
)

_ROADMAP_URL = "https://pve.proxmox.com/wiki/Roadmap"
# Release-history headings on the roadmap read "Proxmox VE X.Y".
_VERSION_RE = re.compile(r"Proxmox VE (\d+)\.(\d+)")
_HTTP_TIMEOUT = 10.0


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _running_version() -> str:
    """Return the cluster's running version string from the Proxmox API."""
    # /version returns e.g. {"version": "8.2.4", "release": "8.2", "repoid": ...}
    data: dict[str, Any] = proxmox_get(lambda: get_client().version.get())
    return str(data.get("version") or data.get("release") or "unknown")


def _major_minor(text: str) -> tuple[int, ...]:
    """Parse the leading numeric components of a version string (major, minor)."""
    parts = re.findall(r"\d+", text or "")
    return tuple(int(p) for p in parts[:2])


def _latest_release() -> tuple[str, str | None]:
    """
    Fetch the latest Proxmox VE release from the roadmap.

    Returns ``(version, None)`` on success or ``("", reason)`` on any failure —
    this is the best-effort boundary and never raises.
    """
    try:
        import httpx

        resp = httpx.get(_ROADMAP_URL, timeout=_HTTP_TIMEOUT, follow_redirects=True)
        resp.raise_for_status()
    except Exception as exc:  # noqa: BLE001 — best-effort network boundary
        logger.warning("Update check: roadmap fetch failed: %s", exc)
        return "", f"could not reach the Proxmox roadmap ({exc})"

    matches = _VERSION_RE.findall(resp.text)
    if not matches:
        return "", "could not parse a version from the Proxmox roadmap page"
    latest = max((int(a), int(b)) for a, b in matches)
    return f"{latest[0]}.{latest[1]}", None


def _check_updates() -> str:
    running_ver = _running_version()
    latest_ver, err = _latest_release()

    lines = ["## Proxmox Update Check\n", f"**Running version:** {running_ver}"]

    if err:
        lines.append(f"**Latest release:** unavailable — {err}")
        lines.append(
            "\nThe running version above is read from your cluster. The "
            "latest-release lookup is best-effort and needs outbound access to "
            f"pve.proxmox.com. Check {_ROADMAP_URL} manually."
        )
        return "\n".join(lines)

    lines.append(f"**Latest release (roadmap):** {latest_ver}")

    running_t = _major_minor(running_ver)
    latest_t = _major_minor(latest_ver)

    if running_t and latest_t and running_t < latest_t:
        lines.append(
            f"\n⬆️ **An upgrade is available** — {running_ver} → {latest_ver}. "
            f"Review the upgrade notes before upgrading: {_ROADMAP_URL}"
        )
    elif running_t and latest_t:
        lines.append(
            f"\n✅ You are on the latest minor release ({latest_ver}). Point "
            f"releases and package updates may still be available — run "
            f"`apt update && apt full-upgrade` on each node to be current."
        )
    else:
        lines.append(
            f"\nCould not compare the versions reliably; check {_ROADMAP_URL}."
        )
    return "\n".join(lines)


# ---------------------------------------------------------------------------
# Tool registration
# ---------------------------------------------------------------------------


def register(mcp: FastMCP) -> None:
    """Register the update-check tool onto the given FastMCP instance."""

    @mcp.tool(
        name="check_proxmox_updates",
        title="Check for Proxmox Updates",
        description=(
            "Check whether the connected Proxmox cluster is running the latest "
            "released version. Reads the running version from the Proxmox API and "
            "compares it to the latest Proxmox VE release published on the "
            "official roadmap (pve.proxmox.com/wiki/Roadmap). Best-effort: if the "
            "roadmap cannot be reached, the running version is still reported."
        ),
        annotations=_READ_ONLY,
    )
    async def check_proxmox_updates() -> str:
        return await safe("check_proxmox_updates", _check_updates)

File: clustr-ts/src/tools/read/updates.ts
/**
 * Read-only tool: is the connected Proxmox cluster running the latest release?
 *
 * Reads the running version from the Proxmox API (/version) and compares it to
 * the latest pve-manager version in Proxmox's APT repository, the structured
 * Packages index apt itself consumes. Track-aware: it maps the running major to
 * its Debian codename and queries that track, so it reports the precise
 * point-release apt would install and never false-positives on an
 * in-development next major.
 *
 * Best-effort: if the package index is unreachable (offline / locked-down
 * network), the running version is still reported.
 */

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { fetch } from "undici";

import { proxmoxGet } from "../../proxmox.js";
import { safe } from "../../safe.js";
import { pendingUpdates } from "./apt.js";

const READ = {
  readOnlyHint: true,
  destructiveHint: false,
  idempotentHint: true,
} as const;

const CODENAMES: Record<number, string> = { 9: "trixie", 8: "bookworm", 7: "bullseye" };
const ROADMAP_URL = "https://pve.proxmox.com/wiki/Roadmap";
const packagesUrl = (codename: string): string =>
  `https://download.proxmox.com/debian/pve/dists/${codename}` +
  `/pve-no-subscription/binary-amd64/Packages`;

function versionTuple(text: string): number[] {
  return (text.match(/\d+/g) ?? []).slice(0, 3).map(Number);
}

function isLess(a: number[], b: number[]): boolean {
  const len = Math.max(a.length, b.length);
  for (let i = 0; i < len; i++) {
    const x = a[i] ?? 0;
    const y = b[i] ?? 0;
    if (x !== y) return x < y;
  }
  return false;
}

/** Highest pve-manager Version: from a Debian Packages index, or null. */
export function parsePveManagerVersion(packagesText: string): string | null {
  let best: number[] | null = null;
  let bestStr: string | null = null;
  for (const stanza of packagesText.split("\n\n")) {
    if (!/^Package: pve-manager$/m.test(stanza)) continue;
    const m = stanza.match(/^Version: (\S+)$/m);
    if (!m) continue;
    const parsed = versionTuple(m[1]);
    if (parsed.length && (best === null || isLess(best, parsed))) {
      best = parsed;
      bestStr = m[1];
    }
  }
  return bestStr;
}

async function latestOnTrack(
  major: number,
): Promise<{ version: string; error: string | null }> {
  const codename = CODENAMES[major];
  if (!codename) {
    return {
      version: "",
      error: `unrecognized Proxmox VE release track (major ${major}); cannot map it to a Debian repository`,
    };
  }
  try {
    const resp = await fetch(packagesUrl(codename), {
      signal: AbortSignal.timeout(15000),
    });
    if (!resp.ok) {
      return { version: "", error: `package index returned HTTP ${resp.status}` };
    }
    const latest = parsePveManagerVersion(await resp.text());
    if (!latest) {
      return {
        version: "",
        error: `could not find pve-manager in the ${codename} package index`,
      };
    }
    return { version: latest, error: null };
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return { version: "", error: `could not reach the Proxmox package index (${msg})` };
  }
}

export function register(server: McpServer): void {
  server.registerTool(
    "check_proxmox_updates",
    {
      title: "Check for Proxmox Updates",
      description:
        "Check for Proxmox updates. Leads with the LOCAL pending-package count " +
        "per node (the Node → Updates panel, instant, no internet from this " +
        "tool) and highlights kernel/PVE updates. Also makes a best-effort " +
        "comparison to the latest release on your track via the public package " +
        "index (skipped cleanly if that can't be reached).",
      annotations: READ,
    },
    async () =>
      safe("check_proxmox_updates", async () => {
        const v = (await proxmoxGet("/version")) as Record<string, any>;
        const runningVer = String(v.version ?? v.release ?? "unknown");
        const runningT = versionTuple(runningVer);
        const lines = ["## Proxmox Update Check\n", `**Running version:** ${runningVer}\n`];

        // Primary, local, reliable: pending APT updates per node.
        const nodes = (await proxmoxGet("/nodes")) as Record<string, any>[];
        lines.push("**Pending package updates (local):**");
        for (const n of nodes) {
          try {
            const { count, notable } = await pendingUpdates(String(n.node));
            if (count === 0) lines.push(`- ${n.node}: ✅ up to date`);
            else
              lines.push(
                `- ${n.node}: ⬆️ ${count} pending` +
                  (notable.length ? ` (incl. ${notable.join(", ")}, reboot likely)` : ""),
              );
          } catch (err) {
            const msg = err instanceof Error ? err.message : String(err);
            lines.push(`- ${n.node}: could not read pending updates (${msg})`);
          }
        }

        // Secondary, best-effort: latest-on-track from the public index.
        if (runningT.length) {
          const { version: latestVer, error } = await latestOnTrack(runningT[0]);
          if (!error && latestVer) {
            lines.push(
              `\n**Latest on your track (pve-no-subscription):** ${latestVer}` +
                (isLess(runningT, versionTuple(latestVer))
                  ? `, ⬆️ newer than your ${runningVer}.`
                  : ", you're current."),
            );
          }
          // If the external lookup failed, we simply omit it. The local count above is the answer.
        }
        lines.push("\nRun `apt update && apt full-upgrade` on a node to apply (read upgrade notes first).");
        return lines.join("\n");
      }),
  );
}
